oap-server/server-core/src/main/java/org/apache/skywalking/oap/server/core/storage/PersistenceTimer.java

```diff
@@ -92,73 +105,153 @@ public void start(ModuleManager moduleManager, CoreModuleConfig moduleConfig) {
+            });
 
-            HistogramMetrics.Timer executeLatencyTimer = executeLatency.createTimer();
-            try {
-                List<List<PrepareRequest>> partitions = Lists.partition(prepareRequests, maxSyncoperationNum);
-                CountDownLatch countDownLatch = new CountDownLatch(partitions.size());
-                for (final List<PrepareRequest> partition : partitions) {
-                    executorService.submit(() -> {
+            List<Future<?>> batchFutures = new ArrayList<>();
+            for (int i = 0; i < syncOperationThreadsNum; i++) {
+                Future<?> batchFuture = executorService.submit(() -> {
+                    // consume the metrics
+                    while (!stop.get()) {
+                        List<PrepareRequest> partition = prepareQueue.pop();
+                        if (partition.isEmpty()) {
+                            break;
+                        }
+                        HistogramMetrics.Timer executeLatencyTimer = executeLatency.createTimer();
                         try {
                             if (CollectionUtils.isNotEmpty(partition)) {
                                 batchDAO.synchronous(partition);
                             }
                         } catch (Throwable e) {
                             log.error(e.getMessage(), e);
                         } finally {
-                            countDownLatch.countDown();
+                            executeLatencyTimer.finish();
                         }
-                    });
-                }
-                countDownLatch.await();
-            } finally {
-                executeLatencyTimer.finish();
+                    }
+                    return null;
+                });
+                batchFutures.add(batchFuture);
+            }
+
+            // Wait for prepare stage is done.
+            prepareStageCountDownLatch.await();
+            prepareQueue.noFurtherAppending();
+            // Wait for batch stage is done.
+            for (Future<?> result : batchFutures) {
+                result.get();
             }
+
         } catch (Throwable e) {
             errorCounter.inc();
             log.error(e.getMessage(), e);
         } finally {
+
             if (log.isDebugEnabled()) {
                 log.debug("Persistence data save finish");
             }
 
-            prepareRequests.clear();
+            stop.set(true);
+            allTimer.finish();
             lastTime = System.currentTimeMillis();
         }
 
         if (debug) {
             log.info("Batch persistence duration: {} ms", System.currentTimeMillis() - startTime);
         }
     }
+
+    @RequiredArgsConstructor
+    private static class BlockingBatchQueue<E> {
+
+        @Getter
+        private final int maxBatchSize;
+
+        @Getter
+        private boolean inAppendingMode = true;
+
+        private final List<E> elementData = new ArrayList<>(50000);
+
+        public void push(List<E> elements) {
```

Comment:
I suggest using the same naming family as JDK's classes, `push` / `pop` for Stacks, `offer` / `poll` for Queues, since this is a queue, let's use `offer` / `poll`, WDYT?

Reply:
ok. got it